### PR TITLE
DescriptionGenerator: map work#related_links to Cocina description

### LINKED ISSUE #341: Map related links to descriptive cocina

### PR DESCRIPTION
**Pull request by jmartin-sul**
## Why was this change made?

closes #341 

## How was this change tested?

unit tests, rails console

## Which documentation and/or configurations were updated?

n/a, hopefully specs cover it

### CODE REVIEW

File: app/services/description_generator.rb
Comment (mjgiarlo) [inline app/services/description_generator.rb:155]: I think `"uri":` here should be `uri:` in line with the changes in https://github.com/sul-dlss/happy-heron/pull/407/files#diff-f21f5ecf9be9f56bd3c574ea7cb992626d29a9a41cc2926f9f643eadf3a15c55 Because these keys are not actually strings; they're symbols. Not sure why Sorbet is tricked by this!

Comment (jmartin-sul) [reply app/services/description_generator.rb:155]: ah, great!  yeah, i thought the `"string":` construction for hash keys was a bit odd, but i figured i'd stay consistent.   mean to ask but forgot, happy to change.

Comment (mjgiarlo) [reply app/services/description_generator.rb:155]: I wonder if some of this could move into a new `RelatedLink#to_h` method, but it's not a big deal. Also, we don't (yet?) do this in other models. This way, each model could take a little more responsibility for its own representation, and the description generator can handle the formatting external to the model? I could go either way on it. Would like to hear what you think, though.

File: spec/services/description_generator_spec.rb
Comment (mjgiarlo) [inline spec/services/description_generator_spec.rb:39]: It might be nice to have an example that shows a RelatedLink with a missing title. Worth making factory & spec tweaks for this?

File: app/services/description_generator.rb
Comment (mjgiarlo) [inline app/services/description_generator.rb:153]: I wonder if some of this could move into a new RelatedLink#to_h method, but it's not a big deal. Also, we don't (yet?) do this in other models. This way, each model could take a little more responsibility for its own representation, and the description generator can handle the formatting external to the model? I could go either way on it. Would like to hear what you think, though.

File: spec/services/description_generator_spec.rb
Comment (jmartin-sul) [reply spec/services/description_generator_spec.rb:39]: good call, will do...

File: app/services/description_generator.rb
Comment (jmartin-sul) [reply app/services/description_generator.rb:153]: i like this idea, though i'd also worry slightly about the `#to_h` representation being very cocina specific... but maybe that's ok?  it does seem like the main use of a model's hash representation would be for translation to cocina representation.  

looking at this example:  https://github.com/sul-dlss-labs/cocina-descriptive-metadata/pull/115/files

i'm not sure that i'd expect such a cocina specific represenation from `#to_h`.

pref for starting that here in this PR, vs doing this one case in a follow on PR, vs updating all of the hash building in this class in one fell swoop in a separate PR?

could also have a `#to_cocina` method instead of using the `#to_h` method.

**Comment by jmartin-sul** [comment]
turned to draft, will fix the output to match the example here after i get back from lunch: https://github.com/sul-dlss-labs/cocina-descriptive-metadata/pull/115/files

File: app/services/description_generator.rb
Comment (mjgiarlo) [reply app/services/description_generator.rb:153]: `#to_cocina` WFM too. Opinions, co-TL, @jcoyne ?

**Comment by jmartin-sul** [comment]
ok, i:
* fixed the weird not-string keys in `#related_resource`
* updated the factory and spec to test for missing link title
* fixed the output to match the structure described in this example:  https://github.com/sul-dlss-labs/cocina-descriptive-metadata/blob/master/h2_cocina_mappings/h2_to_cocina_related_resource.txt

i'll start a branch based on this branch to try out the `#to_h`/`#to_cocina` refactoring (happy to roll that work into this PR if i get that done before this is merged).

rebased on latest `main`, ready for another look, i think.

**Comment by jmartin-sul** [comment]
d'oh, `sig` is now incorrect for the newly added method, turning to draft and fixing...

File: app/services/description_generator.rb
Comment (ndushay) [inline app/services/description_generator.rb:164]: ew.  but we probably have this in other spots.   this is where maybe we could use the references to cocina types?

File: app/services/description_generator.rb
Comment (ndushay) [reply app/services/description_generator.rb:166]: i'd prefer `#to_cocina` or even `#to_cocina_hash` or whatever, but wondering if this refactoring should get priority over other work in ready column ...

**Review by ndushay — APPROVED** [review]

File: app/services/description_generator.rb
Comment (mjgiarlo) [reply app/services/description_generator.rb:166]: Like I said, not a big deal, so not a high priority.

**Review by mjgiarlo — APPROVED** [review]
LGTM, merge when undrafted.

File: app/services/description_generator.rb
Comment (jmartin-sul) [reply app/services/description_generator.rb:164]: yeah, this looks yucky.  i'm about to push something that's... a little longer, but much clearer about the expected structure (imho), so hopefully that'll be less yuck?

**Comment by jmartin-sul** [comment]
@ndushay and @mjgiarlo here's the `sig` i ended up with, which i find clearer, though it is longer:
```ruby
sig do
  returns(T::Array[
    T.any(
      { type: String, access: { url: T::Array[{ value: String }] } },
      { type: String, access: { url: T::Array[{ value: String }] }, title: T::Array[{ value: String }] }
    )
  ])
end
```

**Pull request merged**
